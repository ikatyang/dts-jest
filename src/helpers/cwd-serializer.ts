export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
