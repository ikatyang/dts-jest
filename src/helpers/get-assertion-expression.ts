import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
