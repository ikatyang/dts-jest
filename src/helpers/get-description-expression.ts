import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
