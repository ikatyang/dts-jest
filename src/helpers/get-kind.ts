import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
