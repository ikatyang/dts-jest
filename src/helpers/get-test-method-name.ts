import {TriggerKind} from '../definitions';

export const get_test_method_name = (kind: TriggerKind) => {
  switch (kind) {
    case TriggerKind.None:
    case TriggerKind.Show:
      return 'test';
    case TriggerKind.Only:
      return 'test.only';
    case TriggerKind.Skip:
      return 'test.skip';
    default:
      throw new Error(`Unexpected kind ${JSON.stringify(kind)}`);
  }
};
