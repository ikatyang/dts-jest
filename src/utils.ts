import * as request from 'request';
import * as ts from 'typescript';
import {ServerPage} from './server';

export const traverse_node = (node: ts.Node, callback: (node: ts.Node) => void) => {
  if (node.kind !== ts.SyntaxKind.SourceFile) {
    callback(node);
  }
  ts.forEachChild(node, child => traverse_node(child, callback));
};

export const remove_spaces = (text: string) =>
  text.replace(/^\s+|\s+$|\n/mg, '');

export const defaults = <T>(value: undefined | T, default_value: T): T =>
  (value === undefined)
    ? default_value
    : value;

export const request_server = (port: number, page: ServerPage, args?: {}, callback?: (body: string) => void) => {
  const url = `http://127.0.0.1:${port}${page}`;
  return request.get(
    url,
    {
      qs: args,
    },
    (error, response, body) => {
      // istanbul ignore next
      if (error || response.statusCode !== 200) {
        throw new Error(`Error: url=${url} args=${args} -> ${
          error
            ? error
            : `status=${response.statusCode}`
        }`);
      } else if (callback) {
        callback(body);
      }
    },
  );
};
