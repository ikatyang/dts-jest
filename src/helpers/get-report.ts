import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
