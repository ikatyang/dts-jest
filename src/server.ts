import * as express from 'express';
import * as http from 'http';
import * as request from 'request';
import * as ts from 'typescript';
import {get_debug, get_server_port, get_tsconfig} from './config';
import {package_name, JestConfig, Snapshots, TriggerLines} from './definitions';
import {get_snapshots, request_server} from './utils';

export enum ServerPage {
  Snapshots = '/snapshots',
  Close = '/close',
  PID = '/pid',
  Reset = '/reset',
}

export class Server {

  public app: express.Express;
  public server: http.Server;
  public program: ts.Program;
  public compiler_options: ts.CompilerOptions;

  public last_reset: number = 0;
  public reseting: boolean = true;
  public callbacks: (() => void)[] = [];
  public debug: boolean;

  constructor(config: JestConfig) {
    this.app = express();
    this.init_app();

    const port = get_server_port(config);
    this.server = this.app.listen(port);

    this.debug = get_debug(config);
    this.compiler_options = get_tsconfig(config);

    this.log(`listening on port ${port}`);
  }

  public static request_close(port: number) {
    return request_server(port, ServerPage.Close);
  }

  public static request_reset_mark(port: number) {
    return request_server(port, ServerPage.Reset);
  }

  public static request_reset_start(port: number, filenames: string[]) {
    return request_server(port, ServerPage.Reset, {filename: filenames});
  }

  public static request_pid(port: number, callback: (pid: number) => void) {
    return request_server(port, ServerPage.PID, undefined, body => {
      callback(+body);
    });
  }

  public static request_snapshots(
      port: number,
      filename: string,
      trigger_lines: TriggerLines,
      callback: (snapshots: Snapshots) => void) {
    request_server(
      port,
      ServerPage.Snapshots,
      {
        filename,
        lines: trigger_lines.join(','),
      },
      body => {
        callback(JSON.parse(body));
      },
    );
  }

  public reset_program(filenames: string[]) {
    this.log(`reseting`);
    const old_program = this.program;
    this.callbacks = [];

    const current = Date.now();
    this.reseting = true;
    this.last_reset = current;

    setImmediate(() => {
      const program = ts.createProgram(filenames, this.compiler_options, undefined, old_program);
      if (this.last_reset === current) {
        this.program = program;
        this.reseting = false;
        while (this.callbacks.length !== 0) {
          const callback = this.callbacks.pop();
          callback!();
        }
      }
    });
  }

  public init_app() {
    this.app.get(ServerPage.Snapshots, (request, response) => {
      const filename = (request.query.filename as string);
      const lines = (request.query.lines as string).split(',').map(str => +str);
      const callback = () => {
        const snapshots = get_snapshots(this.program, filename, lines);
        response.send(JSON.stringify(snapshots));
      };
      if (this.reseting) {
        this.callbacks.push(callback);
      } else {
        callback();
      }
    });
    this.app.get(ServerPage.PID, (_, response) => {
      response.send(process.pid.toString());
    });
    this.app.get(ServerPage.Close, () => {
      this.close();
    });
    this.app.get(ServerPage.Reset, request => {
      const filename: undefined | string | string[] = request.query.filename;
      if (filename === undefined) {
        this.reseting = true;
      } else {
        if (typeof filename === 'string') {
          this.reset_program([filename]);
        } else {
          this.reset_program(filename);
        }
      }
    });
  }

  public close() {
    this.log(`closing`);
    this.server.close();
    process.exit(0);
  }

  public log(message: string) {
    if (this.debug) {
      // tslint:disable-next-line:no-console
      console.log(`[${package_name}] ${message}`);
    }
  }

}

export const create_server = (config: JestConfig) =>
  new Server(config);
