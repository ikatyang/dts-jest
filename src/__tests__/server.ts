import {RequestCallback} from 'request';

let body: string;

jest.mock('request', () => ({
  get: (url: string, options: any, callback: RequestCallback) => {
    callback(null, {statusCode: 200} as any, body);
    return [url, options];
  },
}));

jest.mock('express', () => () => ({
  listen: () => ({
    close: jest.fn(),
  }),
  get: jest.fn(),
}));

jest.mock('typescript', () => ({
  createProgram: jest.fn(() => jest.fn()),
}));

jest.mock('../helpers', () => ({
  create_snapshots: jest.fn(),
}));

import * as ts from 'typescript';
import {create_snapshots} from '../helpers';
import {create_server, Server} from '../server';
import {request_server} from '../utils';

const port = 10086;

test('Server.request_close() should request correctly', () => {
  expect(Server.request_close(port)).toMatchSnapshot();
});

test('Server.request_pid() should request correctly', () => {
  body = JSON.stringify(10086);
  expect(Server.request_pid(port, pid => {
    expect(pid).toMatchSnapshot();
  })).toMatchSnapshot();
});

test('Server.request_reset_mark() should request correctly', () => {
  expect(Server.request_reset_mark(port)).toMatchSnapshot();
});

test('Server.request_reset_start() should request correctly', () => {
  expect(Server.request_reset_start(port, ['f1.ts', 'f2.ts'])).toMatchSnapshot();
});

test('Server.request_reset_start() should request correctly', () => {
  body = JSON.stringify({
    1: 'snapshot 1',
    2: 'snapshot 2',
    3: 'snapshot 3',
  });
  expect(Server.request_snapshots(port, 'f1.ts', [1, 2, 3], snapshots => {
    expect(snapshots).toMatchSnapshot();
  })).toMatchSnapshot();
});

describe('server', () => {
  const exit = process.exit;
  const log = console.log;
  beforeAll(() => {
    Object.defineProperty(process, 'exit', {
      value: jest.fn(),
    });
    Object.defineProperty(console, 'log', {
      value: jest.fn(),
    });
  });

  let server: Server;
  beforeEach(() => {
    server = create_server({
      rootDir: 'path/to/somewhere',
      globals: {
        _dts_jest_: {
          tsconfig: {},
        },
      },
    } as any);
  });

  test('#close() should close server and exit process', () => {
    server.close();
    expect(server.server.close).toBeCalled();
    expect(process.exit).toBeCalled();
  });

  test('#log() should log if debug = true', () => {
    Server.prototype.log.call({debug: true}, 'message');
    expect((console.log as any as jest.MockInstance<any>).mock.calls[0][0]).toMatchSnapshot();
  });

  test('#reset_program() should reset program and execute callbacks if exist', done => {
    const origin_program = server.program;
    server.reset_program([]);

    const callback = jest.fn();
    server.callbacks = [callback];

    setImmediate(() => {
      expect(ts.createProgram).toBeCalled();
      expect(callback).toBeCalled();
      expect(server.program).not.toBe(origin_program);
      done();
    });
  });

  test('#reset_program() should not reset program and while it is not the last reseting', done => {
    const origin_program = server.program;
    server.reset_program([]);

    const callback = jest.fn();
    server.callbacks = [callback];

    server.last_reset = Infinity;

    setImmediate(() => {
      expect(ts.createProgram).toBeCalled();
      expect(callback).not.toBeCalled();
      expect(server.program).toBe(origin_program);
      done();
    });
  });

  afterAll(() => {
    Object.defineProperty(console, 'log', {
      value: log,
    });
  });
});

describe('server events', () => {
  type EventReturn = [string, (request: any, response: any) => void];

  const app = {get: (...args: any[]) => args};

  describe('server#init_event_snapshots()', () => {
    const request = {query: {filename: 'path/to/somewhere', lines: '1,2,3'}};

    let response: {send(): void};
    beforeEach(() => {
      response = {send: jest.fn()};
    });

    it('should call callback while reseting is false', () => {
      const server = {app, reseting: false};
      const [event, callback]: EventReturn = Server.prototype.init_event_snapshots.call(server);
      expect(event).toMatchSnapshot();
      callback(request, response);
      expect(response.send).toBeCalled();
      expect(create_snapshots).toBeCalled();
    });
    it('should push callback to callbacks while reseting is true', () => {
      const server = {app, reseting: true, callbacks: []};
      const [event, callback]: EventReturn = Server.prototype.init_event_snapshots.call(server);
      expect(event).toMatchSnapshot();
      callback(request, response);
      expect(response.send).not.toBeCalled();
      expect(server.callbacks.length).toBeGreaterThan(0);
    });
  });

  test('server#init_event_pid() should setup correctly', () => {
    const server = {app};
    const [event, callback]: EventReturn = Server.prototype.init_event_pid.call(server);
    const response = {send: jest.fn()};
    callback(undefined, response);
    expect(event).toMatchSnapshot();
    expect(response.send).toBeCalledWith(process.pid.toString());
  });

  test('server#init_event_close() should setup correctly', () => {
    const server = {app, close: jest.fn()};
    const [event, callback]: EventReturn = Server.prototype.init_event_close.call(server);
    callback(undefined, undefined);
    expect(event).toMatchSnapshot();
    expect(server.close).toBeCalled();
  });

  describe('server#init_event_reset()', () => {
    let server: {app: any, reset_program: any, reseting: any};
    beforeEach(() => {
      server = {app, reset_program: jest.fn(), reseting: false};
    });

    const filename = 'some-file.ts';

    it('should set reseting while filename is undefined', () => {
      const [event, callback]: EventReturn = Server.prototype.init_event_reset.call(server);
      callback({query: {filename: undefined}}, undefined);
      expect(server.reseting).toBe(true);
    });
    it('should reset program while filename is string', () => {
      const [event, callback]: EventReturn = Server.prototype.init_event_reset.call(server);
      callback({query: {filename}}, undefined);
      expect(server.reset_program).toBeCalledWith([filename]);
    });
    it('should reset program while filename is string[]', () => {
      const [event, callback]: EventReturn = Server.prototype.init_event_reset.call(server);
      callback({query: {filename: [filename]}}, undefined);
      expect(server.reset_program).toBeCalledWith([filename]);
    });
  });
});
