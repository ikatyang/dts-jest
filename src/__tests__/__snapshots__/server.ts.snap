// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server.request_close() should request correctly 1`] = `
Array [
  "http://127.0.0.1:10086/close",
  Object {
    "qs": undefined,
  },
]
`;

exports[`Server.request_pid() should request correctly 1`] = `10086`;

exports[`Server.request_pid() should request correctly 2`] = `
Array [
  "http://127.0.0.1:10086/pid",
  Object {
    "qs": undefined,
  },
]
`;

exports[`Server.request_reset_mark() should request correctly 1`] = `
Array [
  "http://127.0.0.1:10086/reset",
  Object {
    "qs": undefined,
  },
]
`;

exports[`Server.request_reset_start() should request correctly 1`] = `
Array [
  "http://127.0.0.1:10086/reset",
  Object {
    "qs": Object {
      "filename": Array [
        "f1.ts",
        "f2.ts",
      ],
    },
  },
]
`;

exports[`Server.request_reset_start() should request correctly 2`] = `
Object {
  "1": "snapshot 1",
  "2": "snapshot 2",
  "3": "snapshot 3",
}
`;

exports[`Server.request_reset_start() should request correctly 3`] = `undefined`;

exports[`server Server#log() should log if debug = true 1`] = `"[dts-jest] message"`;

exports[`server events Server#init_event_close() should setup correctly 1`] = `"/close"`;

exports[`server events Server#init_event_pid() should setup correctly 1`] = `"/pid"`;

exports[`server events Server#init_event_snapshots() should call callback while reseting is false 1`] = `"/snapshots"`;

exports[`server events Server#init_event_snapshots() should push callback to callbacks while reseting is true 1`] = `"/snapshots"`;
